R/qgis-plugin.R

```diff
@@ -0,0 +1,225 @@
+#'
+#' If you installed, removed, enabled or disabled plugins in the QGIS GUI, then
+#' run `qgis_configure()` to make those changes available in R.
+#'
+#' If `names` is not provided to `qgis_enable_plugins()`, it is assumed that all
+#' _disabled_ plugins are to be enabled.
+#' If `names` is not provided to `qgis_disable_plugins()`, it is assumed that all
+#' _enabled_ plugins are to be disabled.
+#' Note that the 'processing' plugin is ignored, because it is always available
+#' to 'qgis_process' (not QGIS though).
+#'
+#' @note
+#' Only plugins that implement Processing providers are supported.
+#' Installing or removing plugins is not supported.
+#'
+#' @inheritParams qgis_run
+#' @param which String defining which plugins to select, based on their
+#' status in QGIS (enabled or disabled).
+#' Must be one of: `"all"`, `"enabled"`, `"disabled"`.
+#' @param names Optional character vector of plugin names.
+#' @return
+#' A tibble of plugins and their status.
+#' @export
+qgis_plugins <- function(
+    which = "all",
+    query = FALSE,
+    quiet = TRUE
+) {
+
+  assert_that(which %in% c("all", "enabled", "disabled"))
+  assert_that(is.flag(query), !is.na(query))
+  assert_that(is.flag(quiet), !is.na(quiet))
+
+  if (query) {
+    qgisprocess_cache$plugins <- qgis_query_plugins(quiet = quiet)
+  }
+
+  plugins <- qgisprocess_cache$plugins
+
+  if (!quiet && query) message(glue(
+    "{ nrow(plugins[plugins$enabled, ]) } out of { nrow(plugins) } ",
+    "available processing provider plugins are enabled."
+  ))
+
+  if (!quiet && !query) message(
+    "Reading plugin list from the qgisprocess cache.\n",
+    "  If you changed plugin availability or status in the QGIS GUI\n",
+    "  since you loaded the qgisprocess package,\n",
+    "  then you must run `qgis_configure()` or reload the package\n",
+    "  to capture these changes."
+  )
+
+  switch(
+    which,
+    "all" = plugins,
+    "enabled" = plugins[plugins$enabled, ],
+    "disabled" = plugins[!plugins$enabled, ]
+  )
+}
+
+
+
+#' @keywords internal
+qgis_query_plugins <- function(quiet = FALSE) {
+
+  if (qgis_use_json_output()) {
+
+    out <- qgis_run(args = c("plugins", "--json"))$stdout
+    pluginlist <- jsonlite::fromJSON(out)$plugins
+    plugins <- tibble::enframe(pluginlist)
+    plugins$value <- unlist(plugins$value, use.names = FALSE)
+    colnames(plugins) <- c("name", "enabled")
+
+  } else {
+
+    lines <- readLines(textConnection(qgis_run("plugins")$stdout))
+    pluginvec <- stringr::str_extract(lines, "^\\*?\\s+\\w+")
+    pluginvec <- pluginvec[!is.na(pluginvec)]
+    plugins <- tibble::tibble(
+      name = stringr::str_match(pluginvec, "^\\*?\\s+(\\w+)")[, 2],
+      enabled = stringr::str_detect(pluginvec, "^\\*")
+    )
+
+  }
+
+  return(plugins)
+}
+
+
+
+
+
+#' @keywords internal
+message_disabled_plugins <-
+  function(
+    plugins,
+    prepend_newline = FALSE,
+    startup = FALSE
+  ) {
```

Comment:
https://style.tidyverse.org/functions.html#long-lines-1

Reply:
Not really my business but I guess this would be more standard. It's what you get when putting `function` on the same line as the function name and pressing Ctrl+Shift+A in RStudio:

```r
message_disabled_plugins <- function(plugins,
                                     prepend_newline = FALSE,
                                     startup = FALSE) {
  # ...
}
```

Reply:
Personally I prefer 

```r
message_disabled_plugins <- function(
    plugins,
    prepend_newline = FALSE,
    startup = FALSE
    ) {
  # ...
}
```

But not sure if that's the package's stylistic convention.

Reply:
Hi @Robinlovelace thanks for the suggestions.

I've learned from Dewey to keep the parentheses separate from the contents within the parentheses, if those need > 1 line. E.g. https://github.com/paleolimbot/qgisprocess/pull/80#discussion_r797191968 (and others in that PR). Since this is the style of the package. At least that's what I understood and I do my best to apply 😊, but always open for improved ideas.

The two chunks of your above posts are indeed the two choices given in the tidyverse style guide for long function definitions; none of them comply with the 'keeping parentheses separate' idea though.

@paleolimbot can you clarify further when to do what?

Reply:
2421d22